### PR TITLE
[Feature] Add API to convert graph to bidirected graph

### PR DESCRIPTION
**Pull request by mufeili**
## Description
See issue #419 . For simplicity, we only handle the graph topology and leave feature part to the users.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- [x] `dgl.to_bidirected` API

### CODE REVIEW

**Comment by VoVAllen** [comment]
1. Shall we support this operation on immutable graph?
2. What will happen if graph is a multi graph?

**Comment by mufeili** [comment]
> 1. Shall we support this operation on immutable graph?

This will return a new graph. For immutable graphs, maybe we can check first if it is readonly and then return a readonly one if so?

> 2. What will happen if graph is a multi graph?

If a graph originally has m edges for i -> j and n edges for j -> i, the new graph will have max(m, n) edges for both i -> j and j -> i. So you will still get a multigraph.

**Comment by yzh119** [comment]
I think such operations should only be applicable to mutable and single graphs.

**Comment by mufeili** [comment]
@VoVAllen @yzh119 In fact currently this does not support immutable graphs . Probably I should also check whether this is a batched graph.

**Comment by jermainewang** [comment]
Since we maintained many different types of graphs in python (mutable, immutable, subgraph, batched_graph, nodeflow), it'll become quite complicated if we want to cover all the cases. Here is a proposal to keep the APIs in `dgl.transform` consistent.
1) The default behavior only transforms graph structure. Additional arguments could be specified to transform features as well.
2) The API is functional, aka a new graph structure is returned.
3) The returned graph is always an immutable graph. Thus, if the transformation happens to be identical and the input graph is also an immutable graph, the input graph could be returned.

Let me know your thoughts.

**Comment by mufeili** [comment]
@jermainewang I think the current implementation already matches the first two points. I'm a bit uncertain about the third point. Is it possible to convert an immutable graph to a mutable graph now? What can be pros and cons with the current system to return a normal graph vs an immutable graph?

**Comment by VoVAllen** [comment]
@jermainewang I'm a bit confused about the third point. The returned graph is identical or not cannot be known beforehand. Therefore user have to do extra check whether this is the input graph after transform.  Maybe always return a new graph and provide some API like `g.is_identical(g2)` can help here?

**Comment by jermainewang** [comment]
> @jermainewang I think the current implementation already matches the first two points. I'm a bit uncertain about the third point. Is it possible to convert an immutable graph to a mutable graph now? What can be pros and cons with the current system to return a normal graph vs an immutable graph?

Immutable graph should be more efficient for most of the operations. The problem is no mutation will be allowed afterwards.

> The returned graph is identical or not cannot be known beforehand. Therefore after transform user have to do extra check whether this is the input graph. It's weird.

The third point tries to ease the implementation. For example, when implementing `to_simple` transformation, if I know that the input graph is already a simple graph (by looking at the flag), I'm safe to return the input graph as the output graph directly without a deep copy.

**Comment by mufeili** [comment]
@jermainewang What if we first check whether the input graph is immutable and then handle the two cases separately?

**Comment by jermainewang** [comment]
> @jermainewang What if we first check whether the input graph is immutable and then handle the two cases separately?

That's fine. You could handle it that way.

**Comment by mufeili** [comment]
@jermainewang The API now returns an immutable graph by default and allows returning mutable graphs if needed.

File: src/graph/graph_apis.cc [hunk 1]
```diff
@@ -572,4 +572,24 @@ DGL_REGISTER_GLOBAL("transform._CAPI_DGLToSimpleGraph")
...
+    *bgptr = GraphOp::ToBidirectedMutableGraph(gptr);
+    GraphHandle bghandle = bgptr;
+    *rv = bghandle;
+  });
+
+DGL_REGISTER_GLOBAL("transform._CAPI_DGLToBidirectedImmutableGraph")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    GraphHandle ghandle = args[0];
+    const GraphInterface *ptr = static_cast<const GraphInterface *>(ghandle);
+    const Graph* gptr = static_cast<const Graph*>(ptr);
```
Comment (VoVAllen) [inline src/graph/graph_apis.cc]: gptr can be null if graph is immutable graph

Comment (mufeili) [reply src/graph/graph_apis.cc]: Should be fixed now.

File: src/graph/graph_op.cc [hunk 2]
```diff
@@ -315,4 +315,74 @@ ImmutableGraph GraphOp::ToSimpleGraph(const GraphInterface* graph) {
...
+    }
+  }
+
+  Graph bg;
+  bg.AddVertices(g->NumVertices());
+  for (dgl_id_t u = 0; u < g->NumVertices(); ++u) {
+    for (dgl_id_t v = u; v < g->NumVertices(); ++v) {
+      const auto new_n_e = std::max(n_e[u][v], n_e[v][u]);
+      if (new_n_e > 0) {
+        IdArray us = IdArray::Empty({new_n_e}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
```
Comment (jermainewang) [inline src/graph/graph_op.cc]: Try use the new routines in `include/dgl/array.h`.

File: src/graph/graph_op.cc [hunk 3]
```diff
@@ -315,4 +315,74 @@ ImmutableGraph GraphOp::ToSimpleGraph(const GraphInterface* graph) {
...
+      const auto new_n_e = std::max(n_e[u][v], n_e[v][u]);
+      for (size_t i = 0; i < new_n_e; ++i) {
+        indices.push_back(v);
+      }
+    }
+    indptr[u+1] = indices.size();
+  }
+  CSRPtr csr(new CSR(g->NumVertices(), indices.size(),
+        indptr.begin(), indices.begin(), RangeIter(0), g->IsMultigraph()));
+  return ImmutableGraph(csr);
```
Comment (jermainewang) [inline src/graph/graph_op.cc]: The immutable graph supports construction from a COO structure. This should be easier and more efficient.

File: tests/compute/test_transform.py [hunk 4]
```diff
@@ -95,9 +95,27 @@ def test_simple_graph():
     eset = set(zip(list(F.asnumpy(src)), list(F.asnumpy(dst))))
     assert eset == set(elist)
 
+def test_bidirected_graph():
+    def _test(in_readonly, out_readonly):
+        elist = [(0, 0), (0, 1), (0, 1), (1, 0), (1, 1), (2, 1), (2, 2), (2, 2)]
+        g = dgl.DGLGraph(elist, readonly=in_readonly)
+        elist.append((1, 2))
+        elist = set(elist)
+        big = dgl.to_bidirected(g, out_readonly)
```
Comment (jermainewang) [inline tests/compute/test_transform.py]: It looks like your `to_bidirected` implicitly convert a multigraph to a simple graph, which might not be the expected behavior. It's generally better to let each operator serve only one purpose, and allow the power of combination to cover rich scenarios.

File: src/graph/graph_op.cc [hunk 5]
```diff
@@ -315,4 +315,74 @@ ImmutableGraph GraphOp::ToSimpleGraph(const GraphInterface* graph) {
   return ImmutableGraph(csr);
 }
 
+Graph GraphOp::ToBidirectedMutableGraph(const GraphInterface* g) {
+  std::unordered_map<int, std::unordered_map<int, int>> n_e;
+  for (dgl_id_t u = 0; u < g->NumVertices(); ++u) {
+    for (const dgl_id_t v : g->SuccVec(u)) {
+      n_e[u][v]++;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mufeili) [reply src/graph/graph_op.cc]: If we have `m` edges for `i->j` and `n` edges for `j->i`, then the current implementation will give `max(m, n)` edges for both direction in the returned graph. The hashmap is just used for a first pass edge counting.

File: tests/compute/test_transform.py (hunk 4 above)
Comment (mufeili) [reply tests/compute/test_transform.py]: No, the returned graph is still a multi-graph.

File: src/graph/graph_op.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply src/graph/graph_op.cc]: If you simply make `m` edges for `j->i` and `n` edges for `i->j`, I think this is what `reverse` intends to do.

File: src/graph/graph_op.cc (hunk 3 above)
Comment (mufeili) [reply src/graph/graph_op.cc]: Done.

File: src/graph/graph_op.cc (hunk 2 above)
Comment (mufeili) [reply src/graph/graph_op.cc]: Done.

**Comment by jermainewang** [comment]
Chatted with @mufeili offline. @mufeili is right about the semantics.

File: python/dgl/transform.py [hunk 6]
```diff
@@ -124,4 +124,28 @@ def to_simple_graph(g):
     newgidx = GraphIndex(_CAPI_DGLToSimpleGraph(g._graph.handle))
     return DGLGraph(newgidx, readonly=True)
 
+def to_bidirected(g, readonly=True):
+    """Convert the graph to a bidirected graph.
+
+    The function generates a new graph with no node/edge feature.
+    If g has m edges for i->j and n edges for j->i, then the
+    returned graph will have max(m, n) edges for both i->j and j->i.
```
Comment (jermainewang) [inline python/dgl/transform.py]: Give some examples in the comment?

**Comment by jermainewang** [comment]
Approved! One last little thing, could you add some examples in the comment about two cases (one simple graph, one multigraph). Otherwise, LGTM and you could go ahead merge it afterwards.

**Comment by mufeili** [comment]
@jermainewang Examples have been added.

**Pull request merged**